File: backend/app/services/liveavatar_lite.py
import logging
from typing import Any, Dict, Optional

import httpx

from app.core.config import get_settings

logger = logging.getLogger(__name__)


class LiveAvatarLiteClient:
    def __init__(self) -> None:
        self.settings = get_settings()

    def _base_url(self) -> str:
        base = (self.settings.liveavatar_base_url or self.settings.liveagent_base_url or '').strip()
        return (base or 'https://api.liveavatar.com').rstrip('/')

    def _api_key(self) -> str:
        return (self.settings.liveavatar_api_key or self.settings.liveagent_api_key or '').strip()

    def _provider_headers(self) -> Dict[str, str]:
        key = self._api_key()
        headers = {'Content-Type': 'application/json', 'accept': 'application/json'}
        if key:
            headers['X-API-KEY'] = key
        return headers

    @staticmethod
    def _session_headers(session_token: str) -> Dict[str, str]:
        return {
            'Content-Type': 'application/json',
            'accept': 'application/json',
            'authorization': f'Bearer {session_token}',
        }

    async def create_session_token(
        self,
        *,
        avatar_id: str,
        voice_id: Optional[str] = None,
        context_id: Optional[str] = None,
        language: str = 'en',
        video_encoding: str = 'VP8',
        video_quality: str = 'high',
        is_sandbox: bool = False,
        livekit_config: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        if not self._api_key():
            return {'ok': False, 'error': 'LIVEAVATAR_API_KEY not configured'}

        payload: Dict[str, Any] = {
            'mode': 'LITE',
            'avatar_id': avatar_id,
            'avatar_persona': {'language': language or 'en'},
            'is_sandbox': bool(is_sandbox),
            'video_settings': {'encoding': video_encoding, 'quality': video_quality},
        }
        if voice_id:
            payload['avatar_persona']['voice_id'] = voice_id
        if context_id:
            payload['avatar_persona']['context_id'] = context_id
        if livekit_config:
            payload['livekit_config'] = livekit_config

        url = f'{self._base_url()}/v1/sessions/token'
        try:
            async with httpx.AsyncClient(timeout=httpx.Timeout(20.0)) as client:
                response = await client.post(url, json=payload, headers=self._provider_headers())
                response.raise_for_status()
                raw = response.json()
        except httpx.HTTPStatusError as exc:
            logger.error('LiveAvatar LITE token HTTP error: %s', exc.response.status_code)
            return {'ok': False, 'error': f'HTTP {exc.response.status_code}: token creation failed'}
        except httpx.RequestError as exc:
            logger.error('LiveAvatar LITE token request failed: %s', exc)
            return {'ok': False, 'error': 'token request failed'}

        code = raw.get('code') if isinstance(raw, dict) else None
        data = raw.get('data') if isinstance(raw, dict) else None
        session_id = data.get('session_id') if isinstance(data, dict) else None
        session_token = data.get('session_token') if isinstance(data, dict) else None
        if code != 1000 or not session_id or not session_token:
            return {'ok': False, 'error': f'Provider token creation failed (code={code})', 'raw': raw}

        return {'ok': True, 'session_id': session_id, 'session_token': session_token, 'raw': raw}

    async def start_session(self, *, session_token: str) -> Dict[str, Any]:
        url = f'{self._base_url()}/v1/sessions/start'
        try:
            async with httpx.AsyncClient(timeout=httpx.Timeout(20.0)) as client:
                response = await client.post(url, json={}, headers=self._session_headers(session_token))
                response.raise_for_status()
                raw = response.json()
        except httpx.HTTPStatusError as exc:
            logger.error('LiveAvatar LITE start HTTP error: %s', exc.response.status_code)
            return {'ok': False, 'error': f'HTTP {exc.response.status_code}: session start failed'}
        except httpx.RequestError as exc:
            logger.error('LiveAvatar LITE start request failed: %s', exc)
            return {'ok': False, 'error': 'session start request failed'}

        code = raw.get('code') if isinstance(raw, dict) else None
        data = raw.get('data') if isinstance(raw, dict) else None
        if code != 1000 or not isinstance(data, dict):
            return {'ok': False, 'error': f'Provider session start failed (code={code})', 'raw': raw}

        return {
            'ok': True,
            'session_id': data.get('session_id'),
            'livekit_url': data.get('livekit_url'),
            'livekit_client_token': data.get('livekit_client_token'),
            'livekit_agent_token': data.get('livekit_agent_token'),
            'ws_url': data.get('ws_url'),
            'max_session_duration': data.get('max_session_duration'),
            'raw': raw,
        }

    async def stop_session(self, *, session_id: str, session_token: str) -> Dict[str, Any]:
        urls = [
            f'{self._base_url()}/v1/sessions/stop',
            f'{self._base_url()}/v1/sessions/{session_id}/stop',
        ]
        async with httpx.AsyncClient(timeout=httpx.Timeout(15.0)) as client:
            last_error = 'unknown'
            for url in urls:
                payload = {'session_id': session_id} if url.endswith('/sessions/stop') else {}
                try:
                    response = await client.post(url, json=payload, headers=self._session_headers(session_token))
                    response.raise_for_status()
                    return {'ok': True, 'raw': response.json() if response.text else {}}
                except httpx.HTTPStatusError as exc:
                    last_error = f'HTTP {exc.response.status_code} at {url}'
                    if exc.response.status_code in (404, 405):
                        continue
                    return {'ok': False, 'error': last_error}
                except httpx.RequestError as exc:
                    last_error = str(exc)
                    continue
        return {'ok': False, 'error': f'Unable to stop session. Last error: {last_error}'}

File: frontend/src/liveavatarLite.ts
export interface LiveAvatarLiteSessionStart {
  ok: boolean;
  session_id?: string;
  session_token?: string;
  livekit_url?: string;
  livekit_client_token?: string;
  ws_url?: string;
  agent_ws_registered?: boolean;
  error?: string;
}

export interface LiveAvatarLiteSessionCreate {
  ok: boolean;
  session_id?: string;
  session_token?: string;
  error?: string;
}

interface LiveAvatarLiteSessionStopResponse {
  ok: boolean;
  error?: string;
}

interface LiveAvatarLiteStatusResponse {
  exists?: boolean;
  ready?: boolean;
  ws_connected?: boolean;
  session_state?: string;
  last_error?: string | null;
}

async function requestJson<T>(url: string, init: RequestInit): Promise<T> {
  const response = await fetch(url, init);
  const text = await response.text();
  const payload = text ? safeJsonParse(text) : {};
  if (!response.ok) {
    const detail =
      payload && typeof payload === 'object' && 'detail' in payload
        ? String((payload as Record<string, unknown>).detail || '')
        : '';
    throw new Error(detail || `Request failed: ${response.status}`);
  }
  return payload as T;
}

export async function createLiveAvatarLiteSession(baseUrl: string): Promise<LiveAvatarLiteSessionCreate> {
  const response = await fetch(`${baseUrl}/api/liveavatar/lite/create`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      language: 'en',
      video_quality: 'high',
      video_encoding: 'VP8',
      is_sandbox: false,
    }),
  });
  const text = await response.text();
  const payload = text ? safeJsonParse(text) : {};
  const raw = payload as Record<string, unknown>;

  if (!response.ok) {
    const detail = typeof raw.detail === 'string' ? raw.detail : '';
    const error = typeof raw.error === 'string' ? raw.error : '';
    return { ok: false, error: detail || error || `Request failed: ${response.status}` };
  }

  if (typeof raw.session_id === 'string' && typeof raw.session_token === 'string') {
    return {
      ok: true,
      session_id: raw.session_id,
      session_token: raw.session_token,
    };
  }

  return { ok: false, error: 'Unexpected liveavatar create response shape' };
}

export async function startLiveAvatarLiteSession(baseUrl: string, payload: { session_token: string }): Promise<LiveAvatarLiteSessionStart> {
  const response = await fetch(`${baseUrl}/api/liveavatar/lite/start`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  const text = await response.text();
  const parsed = text ? safeJsonParse(text) : {};
  const raw = parsed as Record<string, unknown>;

  if (!response.ok) {
    const detail = typeof raw.detail === 'string' ? raw.detail : '';
    const error = typeof raw.error === 'string' ? raw.error : '';
    return { ok: false, error: detail || error || `Request failed: ${response.status}` };
  }

  if (
    typeof raw.session_id === 'string' &&
    typeof raw.livekit_url === 'string' &&
    typeof raw.livekit_client_token === 'string'
  ) {
    return {
      ok: true,
      session_id: raw.session_id,
      session_token: typeof raw.session_token === 'string' ? raw.session_token : payload.session_token,
      livekit_url: raw.livekit_url,
      livekit_client_token: raw.livekit_client_token,
      ws_url: typeof raw.ws_url === 'string' ? raw.ws_url : undefined,
      agent_ws_registered: Boolean(raw.agent_ws_registered),
    };
  }

  return { ok: false, error: 'Unexpected liveavatar start response shape' };
}

// Optional fallback/debug helper for passthrough endpoint.
export async function createAndStartLiveAvatarLiteSessionViaNew(baseUrl: string): Promise<LiveAvatarLiteSessionStart> {
  const response = await fetch(`${baseUrl}/api/liveavatar/lite/new`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      language: 'en',
      video_quality: 'high',
      video_encoding: 'VP8',
      is_sandbox: false,
    }),
  });
  const text = await response.text();
  const parsed = text ? safeJsonParse(text) : {};
  const raw = parsed as Record<string, unknown>;

  if (!response.ok) {
    const detail = typeof raw.detail === 'string' ? raw.detail : '';
    const error = typeof raw.error === 'string' ? raw.error : '';
    return { ok: false, error: detail || error || `Request failed: ${response.status}` };
  }

  // HeyGen raw response shape from /api/liveavatar/lite/new passthrough.
  const data = raw.data && typeof raw.data === 'object' ? (raw.data as Record<string, unknown>) : null;
  if (Number(raw.code) === 100 && data && typeof data.session_id === 'string' && typeof data.url === 'string' && typeof data.access_token === 'string') {
    return {
      ok: true,
      session_id: data.session_id,
      session_token: typeof data.session_token === 'string' ? data.session_token : undefined,
      livekit_url: data.url,
      livekit_client_token: data.access_token,
      ws_url: typeof data.ws_url === 'string' ? data.ws_url : undefined,
    };
  }

  return { ok: false, error: 'Unexpected liveavatar /lite/new response shape' };
}

export async function stopLiveAvatarLiteSession(
  baseUrl: string,
  payload: { session_id: string; session_token: string },
): Promise<LiveAvatarLiteSessionStopResponse> {
  return requestJson<LiveAvatarLiteSessionStopResponse>(`${baseUrl}/api/liveavatar/lite/stop`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export async function getLiveAvatarLiteStatus(baseUrl: string, sessionId: string): Promise<LiveAvatarLiteStatusResponse> {
  return requestJson<LiveAvatarLiteStatusResponse>(`${baseUrl}/api/liveavatar/lite/status/${encodeURIComponent(sessionId)}`, {
    method: 'GET',
  });
}

function safeJsonParse(text: string): unknown {
  try {
    return JSON.parse(text);
  } catch {
    return { detail: text };
  }
}

